module/util/context.go

```diff
@@ -0,0 +1,102 @@
+package util
+
+import (
+	"context"
+	"errors"
+	"os"
+	"time"
+)
+
+var ErrChannelClosed = errors.New("channel closed")
+var ErrSignalReceived = errors.New("signal received")
+
+// WithDone wraps a signal channel with a context, and cancels the context when the channel is closed.
+// When the context is Done, the ctx.Err() will either be ErrChannelClosed if the channel closed first,
+// or the error from the underlying context (Canceled, DeadlineExceeded, etc).
+func WithDone(parent context.Context, done <-chan struct{}) (context.Context, context.CancelFunc) {
+	ctx, cancel := context.WithCancel(parent)
+	c := &doneCtx{cancelCtx: ctx}
+	go func() {
+		select {
+		case <-done:
+			cancel()
+			c.err = ErrChannelClosed
+		case <-ctx.Done():
+			c.err = ctx.Err()
+		}
+	}()
+	return ctx, cancel
+}
+
+type doneCtx struct {
+	cancelCtx context.Context
+	err       error
+}
+
+func (d *doneCtx) Done() <-chan struct{} {
+	return d.cancelCtx.Done()
+}
+
+func (d *doneCtx) Err() error {
+	return d.err
+}
+
+func (d *doneCtx) Value(key interface{}) interface{} {
+	return d.cancelCtx.Value(key)
+}
+
+func (d *doneCtx) Deadline() (deadline time.Time, ok bool) {
+	return d.cancelCtx.Deadline()
+}
+
+// WithSignal wraps an os.Signal channel with a context, and calls the cancel a signal is received.
+// When the context is Done, the ctx.Err() will either be ErrSignalReceived if a signal was received,
+// or the error from the underlying context (Canceled, DeadlineExceeded, etc).
+//
+// Note that the context can only be cancelled once and this can only communicate a single signal,
+// even though the channel can receive multiple. This is consistent with context paradigm of closing
+// the done channel on cancel. If you need to check for multiple signals, simply create a new context
+// using WithSignal and the same signal channel, and the new context will handle the next signal on the
+// channel or received later.
+// The specific signal received is available from the Signal() method.
+func WithSignal(parent context.Context, signalChan <-chan os.Signal) (context.Context, context.CancelFunc) {
```

Comment:
I think it makes sense to pass in a variadic list of Signals here, and the signal channel can be created and registered inside this function.

Reply:
the issue is that you may want to accept multiple signals (e.g. one SIGINT to initiate shutdown, and another to abort clean shutdown like in node run). If the channel lives inside a single use context, we lose subsequent events.

Reply:
Actually, looks like golang already has something for this: https://pkg.go.dev/os/signal#NotifyContext

We should look into using that instead.

Reply:
Native `NotifyContext` for golang added in `go 1.16`, however, we are [still in `go 1.15`](https://github.com/onflow/flow-go/blob/master/go.mod#L3), so for now we are good.

Reply:
now that's we're on 1.16, I switched to over to `NotifyContext`. 2e555f6